fix(proguard): Remap method if there's no line number

loewenheim opened the pull request:
Right now, the logic for remapping a frame is:
* Try to remap the frame with line number if available.
* Otherwise try to remap the frame with params.
* Otherwise don't remap the frame.

But even putting aside the matter of params, this is not [what the Python implementation does](https://github.com/getsentry/sentry/blob/11e254e81180b10de90739f838dccee96fb62bb5/src/sentry/lang/java/plugin.py#L102). It uses 0 as the line number if it's missing.

Therefore, we now do:
* Try to remap the frame with line number if available.
* Otherwise try to remap the frame with params.
* Otherwise try to remap the frame with line number 0.

I'm honestly surprised that remapping a frame with line 0 ever works, but I've seen it happen.

Swatinem reviewed (approved):
I believe I even added an explicit fn to the proguard mapper at one point to try remapping unambiguous method names which does not require a line number.

The `0` fallback is indeed surprising TBH.

loewenheim commented:
I updated this to use `ProguardMapper::remap_method` instead of line 0. I also refactored `map_frame` into a more functional version with less repetition (in a separate commit), but I may have gone a bit overboard with the galaxy brain style.
